DateTimeFormat - restructure constructor parameters

hamishwillee opened the pull request:
This follows on from #11501

This is just the first step - restructure the constructor to use subfeatures for parameters, and subfeatures of those for options.

I am in no way sure this is "correct" so this is a first step.

Following on from this I will need to add the new values for **timeZoneName** added in FF91 - see https://bugzilla.mozilla.org/show_bug.cgi?id=1710429.

hamishwillee commented on javascript/builtins/intl/DateTimeFormat.json:
```diff
@@ -113,197 +113,55 @@
-                "support": {
-                  "chrome": {
-                    "version_added": "73"
+                    "version_added": "24"
```
Note, this just copies the parent compat info - assumption is that this was added in first version

hamishwillee commented on javascript/builtins/intl/DateTimeFormat.json:
```diff
@@ -313,9 +171,9 @@
+            "options_parameter": {
               "__compat": {
-                "description": "IANA time zone names in <code>timeZone</code> option",
+                "description": "<code>options</code> parameter",
```
Note, this just copies the parent compat info - assumption is that this was added in first version

hamishwillee commented on javascript/builtins/intl/DateTimeFormat.json:
```diff
@@ -324,19 +182,19 @@
                   },
                   "nodejs": {
-                    "version_added": "4.0.0"
+                    "version_added": "0.12.0"
```
FYI slight exception to "copying the parent compatibility". This copies the parent first introduction when things were "partial". This must be correct because there are options that have versions before v13.

hamishwillee commented on javascript/builtins/intl/DateTimeFormat.json:
```diff
@@ -354,63 +212,370 @@
-                  "webview_android": {
-                    "version_added": "76"
+              },
+              "options_dateStyle_parameter": {
```
All of these are just renamed parameters and moved to subfeatures of the option.

hamishwillee commented on javascript/builtins/intl/DateTimeFormat.json:
```diff
@@ -354,63 +212,370 @@
+                  }
+                }
+              },
+              "options_timeZone_parameter": {
```
So this creates a timeZone option and adds the iana stuff as a subfeature - both with same compat. I think this is probably wrong - timezone came first. Looking at versions, perhaps timeZone should be added as earliest version for the few cases where it is not already the earliest version?

Elchi3 replied:
> Looking at versions, perhaps timeZone should be added as earliest version for the few cases where it is not already the earliest version?

Yeah, this seems more plausible to me, too

hamishwillee replied:
thanks!

hamishwillee commented:
The next step is to add options_timeZoneName_parameter.
1. My inference is that I create options_timeZoneName_parameter with a subfeature options_timeZoneName_parameter_ana to mirror the timeZone care - right?
2. What do I put for the versions in `options_timeZoneName_parameter` - do I just assume this option was introduced at the earliest versions of the constructor?
3. Can you help me work out what version of v8 the new options went into?

Elchi3 commented:
> * My inference is that I create options_timeZoneName_parameter with a subfeature options_timeZoneName_parameter_ana to mirror the timeZone care - right?

This works, I wonder if we're nesting too deep at this point. I think it wouldn't hurt to have it at the same level. I have no strong feeling.

> * What do I put for the versions in `options_timeZoneName_parameter` - do I just assume this option was introduced at the earliest versions of the constructor?

I think that makes sense, yes.

> * Can you help me work out what version of v8 the new options went into?

I can't unfortunately.

Elchi3 commented on javascript/builtins/intl/DateTimeFormat.json:
```diff
@@ -354,63 +212,370 @@
+              },
+              "options_dateStyle_parameter": {
+                "__compat": {
+                  "description": "<code>options_dateStyle</code> parameter",
```
```suggestion
                  "description": "<code>options.dateStyle</code> parameter",
```

hamishwillee replied:
> Given options is an object, a dot feels better to me.

@Elchi3 Thanks, I misread the guidance for the description. Will accept all these.

Elchi3 commented on javascript/builtins/intl/DateTimeFormat.json:
```diff
@@ -354,63 +212,370 @@
+              },
+              "options_dayPeriod_parameter": {
+                "__compat": {
+                  "description": "<code>options_dayPeriod</code> parameter",
```
```suggestion
                  "description": "<code>options.dayPeriod</code> parameter",
```

Elchi3 commented on javascript/builtins/intl/DateTimeFormat.json:
```diff
@@ -354,63 +212,370 @@
+              },
+              "options_fractionalSecondDigits_parameter": {
+                "__compat": {
+                  "description": "<code>options_fractionalSecondDigits</code> parameter",
```
```suggestion
                  "description": "<code>options.fractionalSecondDigits</code> parameter",
```

Elchi3 commented on javascript/builtins/intl/DateTimeFormat.json:
```diff
@@ -354,63 +212,370 @@
+              },
+              "options_hourCycle_parameter": {
+                "__compat": {
+                  "description": "<code>options_hourCycle</code> parameter",
```
```suggestion
                  "description": "<code>options.hourCycle</code> parameter",
```

Elchi3 commented on javascript/builtins/intl/DateTimeFormat.json:
```diff
@@ -354,63 +212,370 @@
+              },
+              "options_timeStyle_parameter": {
+                "__compat": {
+                  "description": "<code>options_timeStyle</code> parameter",
```
```suggestion
                  "description": "<code>options.timeStyle</code> parameter",
```

Elchi3 commented on javascript/builtins/intl/DateTimeFormat.json:
```diff
@@ -354,63 +212,370 @@
+              },
+              "options_timeZone_parameter": {
+                "__compat": {
+                  "description": "<code>options_timeZone</code> parameter",
```
```suggestion
                  "description": "<code>options.timeZone</code> parameter",
```

Elchi3 commented on javascript/builtins/intl/DateTimeFormat.json:
```diff
@@ -354,63 +212,370 @@
-                  "deprecated": false
+                "options_timeZone_parameter_iana": {
+                  "__compat": {
+                    "description": "IANA time zone names in <code>options_timeZone</code> option",
```
```suggestion
                    "description": "IANA time zone names in <code>options.timeZone</code> option",
```

Elchi3 reviewed (changes requested):
Given `options` is an object, a dot feels better to me.

hamishwillee commented:
@Elchi3 Thanks for your review and advice: https://github.com/mdn/browser-compat-data/pull/11544#issuecomment-882407178

I have merged your suggestions, and also updated the timeZone versions to match the constructor, where they didn't already.

Lastly I have added the timeZoneName option and iana subfeatures, mirroring those for timeZone. That makes sense because whether or not it is too deeply nested, we should be consistent. All other browsers are marked false for the new feature - though I only tested on latest Chrome.

I think ready for rereview.

Elchi3 reviewed (approved):
Thanks Hamish! 👍

ddbeck commented:
This PR affected the following features:

## Removed

- `javascript.builtins.Intl.DateTimeFormat.DateTimeFormat.dateStyle`
- `javascript.builtins.Intl.DateTimeFormat.DateTimeFormat.dayPeriod`
- `javascript.builtins.Intl.DateTimeFormat.DateTimeFormat.fractionalSecondDigits`
- `javascript.builtins.Intl.DateTimeFormat.DateTimeFormat.hourCycle`
- `javascript.builtins.Intl.DateTimeFormat.DateTimeFormat.iana_time_zone_names`
- `javascript.builtins.Intl.DateTimeFormat.DateTimeFormat.timeStyle`

## Added

- `javascript.builtins.Intl.DateTimeFormat.DateTimeFormat.locales_parameter`
- `javascript.builtins.Intl.DateTimeFormat.DateTimeFormat.options_parameter`
- `javascript.builtins.Intl.DateTimeFormat.DateTimeFormat.options_parameter.options_dateStyle_parameter`
- `javascript.builtins.Intl.DateTimeFormat.DateTimeFormat.options_parameter.options_dayPeriod_parameter`
- `javascript.builtins.Intl.DateTimeFormat.DateTimeFormat.options_parameter.options_fractionalSecondDigits_parameter`
- `javascript.builtins.Intl.DateTimeFormat.DateTimeFormat.options_parameter.options_hourCycle_parameter`
- `javascript.builtins.Intl.DateTimeFormat.DateTimeFormat.options_parameter.options_timeStyle_parameter`
- `javascript.builtins.Intl.DateTimeFormat.DateTimeFormat.options_parameter.options_timeZone_parameter`
- `javascript.builtins.Intl.DateTimeFormat.DateTimeFormat.options_parameter.options_timeZone_parameter.options_timeZone_parameter_iana`
- `javascript.builtins.Intl.DateTimeFormat.DateTimeFormat.options_parameter.options_timeZoneName_parameter`
- `javascript.builtins.Intl.DateTimeFormat.DateTimeFormat.options_parameter.options_timeZoneName_parameter.options_timeZoneName_parameter_iana`
